Emby.Server.Implementations/ScheduledTasks/Tasks/OptimizeDatabase.cs

```diff
@@ -0,0 +1,100 @@
+using Emby.Server.Implementations.Data;
+using Jellyfin.Data.Entities;
+using MediaBrowser.Controller;
+using MediaBrowser.Common.Configuration;
+using MediaBrowser.Model.Globalization;
+using MediaBrowser.Model.IO;
+using MediaBrowser.Model.Tasks;
+using Microsoft.Extensions.Logging;
+using SQLitePCL.pretty;
+
+namespace Emby.Server.Implementations.ScheduledTasks.Tasks
+{
+    /// <summary>
+    /// Optimizes Jellyfin's database by issuing a VACUUM command.
+    /// </summary>
+    public class OptimizeDatabaseTask : IScheduledTask, IConfigurableScheduledTask
+    {
+        private readonly ILogger<OptimizeDatabaseTask> _logger;
+        private readonly ILocalizationManager _localization;
+        private readonly IServerApplicationPaths _paths;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OptimizeDatabaseTask" /> class.
+        /// </summary>
+        public OptimizeDatabaseTask(
+            ILogger<OptimizeDatabaseTask> logger,
+            ILocalizationManager localization,
+            IServerApplicationPaths paths)
+        {
+            _logger = logger;
+            _localization = localization;
+            _paths = paths;
+        }
+
+        /// <inheritdoc />
+        public string Name => _localization.GetLocalizedString("TaskOptimizeDatabase");
+
+        /// <inheritdoc />
+        public string Description => _localization.GetLocalizedString("TaskOptimizeDatabaseDescription");
+
+        /// <inheritdoc />
+        public string Category => _localization.GetLocalizedString("TasksMaintenanceCategory");
+
+        /// <inheritdoc />
+        public string Key => "OptimizeDatabaseTask";
+
+        /// <inheritdoc />
+        public bool IsHidden => false;
+
+        /// <inheritdoc />
+        public bool IsEnabled => true;
+
+        /// <inheritdoc />
+        public bool IsLogged => true;
+
+        /// <summary>
+        /// Creates the triggers that define when the task will run.
+        /// </summary>
+        /// <returns>IEnumerable{BaseTaskTrigger}.</returns>
+        public IEnumerable<TaskTriggerInfo> GetDefaultTriggers()
+        {
+            return new[]
+            {
+                // Every so often
+                new TaskTriggerInfo { Type = TaskTriggerInfo.TriggerInterval, IntervalTicks = TimeSpan.FromHours(24).Ticks }
+            };
+        }
+
+        /// <summary>
+        /// Returns the task to be executed.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <param name="progress">The progress.</param>
+        /// <returns>Task.</returns>
+        public Task Execute(CancellationToken cancellationToken, IProgress<double> progress)
+        {
+            var dataPath = _paths.DataPath;
+            _logger.LogInformation("Issuing VACUUM command to jellyfin.db...");
+
+            try
+            {
+                using var connection = SQLite3.Open(Path.Combine(dataPath, "jellyfin.db"));
```

Comment:
Should also vacuum `library.db` and `authentication.db`, as those were what was vacuumed, not `jellyfin.db`

Reply:
I know but I don't think it makes much sense given that those databases are going to be deprecated sooner than later (once they're EFCore as well). 

What the rest of you guys think? @jellyfin/server

Reply:
Maybe you could add an option "de-fragment database on startup". So users, who have their database on suboptimal network storage, could disable this or only enable it once in a while if the database is getting too fragmented. Also a user might want to disable this simply to conserve SSD / flash drive durability. Personally I run a secondary dev/beta instance to test out new versions and sometimes deploy a couple of times a day to that instance. Writing 700 Mb every time the server starts adds up to a lot of wasted TBW pretty quickly. In my opinion such an option would please everybody.

Reply:
@ggogel While I understand your concerns, I don't think it's worth it for a number of reasons:

• As we already discussed in Matrix, this is an stopgap while we get a proper database schema. Once we have it, fully integrated with EFCore, we would be able to completely vacuum everything at runtime with an scheduled tasks, which users can configure. So the problem will not be there forever, will be solved at some point. **Imo, this is just an stopgap for another stopgap** and stopgaps are cool but they shouldn't be abused.

• The use case you're exposing is pushing SQLite's design to it's limits already, as it's not designed to run over networks. Same for Jellyfin and it's data. The argument still won't fix the already edge case.


• Modern SSDs aren't that affected by TBW.

•VACUUM actually reduces the file size of the writes.


• Documenting, testing + everything else this new setting... For how many people really? Vacuum at startup has been running for almost a year and you are the first to notice it :).

• Jellyfin is a long running task: as a server, most people have it running in the background. As a long-running task, startups are not frequent.

I also suspect that, given you're running everything over a network, database fragmentation is even worse, that's why you're also experiencing high IO as iirc you mentioned correctly.

TLDR: There's no point imo to add an stopgap for another stopgap which doesn't provide a benefit/solution to a good amount of users. Your usage case is at the edge already.